### PR TITLE
release-4.0.0 (HDS-2191) fix button styles across

### PR DESCRIPTION
**Pull request by mrTuomoK**
## Description

Renews the old Button component and makes changes accordingly across the rest of the components, adds possibility to theme the Button and changes some props to enums. Also adds the possibility to use icons on components by giving them correct classes (what was wanted for buttons, not to contain the spans (example on core side) )

Demos
[react](https://city-of-helsinki.github.io/hds-demo/preview_1251/storybook/react)
[core](https://city-of-helsinki.github.io/hds-demo/preview_1251/storybook/core)
[site](https://city-of-helsinki.github.io/hds-demo/preview_1251/)

You can also check the other components too which use Button-component

## Related Issues

[HDS-2191](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2191)
[HDS-1117](https://helsinkisolutionoffice.atlassian.net/browse/HDS-1117)
[HDS-2195](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2195)
[HDS-2208](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2208)
[HDS-2209](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2209)
[HDS-2184](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2184)
[HDS-1831](https://helsinkisolutionoffice.atlassian.net/browse/HDS-1831)
[HDS-2024](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2024)

## How Has This Been Tested?

- local machine

## Add to changelog
- [x] Added needed line to changelog 

[HDS-2191]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2191?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-1117]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-1117?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-2195]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2195?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-2208]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2208?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-2209]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2209?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-2184]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2184?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HDS-1831]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-1831?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/react/src/components/cookieConsent/languageSwitcher/LanguageSwitcherItem/LanguageSwitcherItem.module.scss [hunk 1]
```diff
@@ -1,11 +1,12 @@
 @import '../../../../styles/common.scss';
-@import '../../../button/button.common.scss';
 @value x-small-down, small-down, small-only, medium-only, medium-up, large-only, large-up, x-large-only, x-large-up from "../../../../styles/breakpoints.scss";
 
 %item {
+  /*
```
Comment (Riippi) [inline packages/react/src/components/cookieConsent/languageSwitcher/LanguageSwitcherItem/LanguageSwitcherItem.module.scss:5]: Some commented lines? Also more commented lines later in this file.

File: site/src/docs/components/stepper/code.mdx [hunk 2]
```diff
@@ -239,15 +239,15 @@ import { StepState, Stepper, Button, IconArrowLeft, IconArrowRight } from 'hds-r
           variant="secondary"
           onClick={() => dispatch({ type: 'setActive', payload: state.activeStepIndex - 1 })}
           style={{ height: 'fit-content', width: 'fit-content' }}
-          iconLeft={<IconArrowLeft />}
+          IconStart={<IconArrowLeft />}
```
Comment (Riippi) [inline site/src/docs/components/stepper/code.mdx:242]: Suddenly IconStart with capped first letter?  Same on three occasions later in this file.

**Comment by Riippi** [comment]
Padding seems weird in these react storybook button examples:
 - with icons
 - Loading
 - Small
 
 Also buttons in datepicker and accordion have odd paddings on buttons when focused.
 
 In Core storybook buttons look normal.

**Review by Riippi — CHANGES_REQUESTED** [review]

File: site/src/docs/components/stepper/code.mdx (hunk 2 above)
Comment (mrTuomoK) [reply site/src/docs/components/stepper/code.mdx:242]: Good spotting

**Comment by mrTuomoK** [comment]
> Padding seems weird in these react storybook button examples:
> 
> * with icons
> * Loading
> * Small
> 
> Also buttons in datepicker and accordion have odd paddings on buttons when focused.
> 
> In Core storybook buttons look normal.

Hmm. Everything is according to designs, could you @SillanmakiA check them (check also Accordion)? I need to check the Datepicker styles from Figma and fix them.

File: CHANGELOG.md [hunk 3]
```diff
@@ -56,6 +58,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Changed
 
 - [Component] What has been changed
+- [Button] Renewed Button component with theming support
 
 #### Fixed
```
Comment (NikoHelle) [inline CHANGELOG.md:64]: Changes to Accordion props (size enums) should also be listed? Are there other affected props in other components?

File: packages/react/src/components/button/Button.tsx [hunk 4]
```diff
@@ -114,25 +155,26 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
...
         )}
         onClick={isLoading ? loadingOnClick : onClick}
         {...rest}
       >
-        {isLoading ? <LoadingSpinner small /> : iconElementLeft}
-        <span className={styles.label}>{isLoading ? loadingText : children}</span>
-        {isLoading ? null : iconElementRight}
+        {isLoading ? <LoadingSpinner small /> : iconElementStart}
+        <span>{isLoading ? loadingText : children}</span>
+        {isLoading ? null : iconElementEnd}
```
Comment (NikoHelle) [inline packages/react/src/components/button/Button.tsx:177]: Side note: Opened a discussion about these old props in Slack.

**Review by NikoHelle — CHANGES_REQUESTED** [review]
This makes buttons some much clearer! Excellent work! Left some comments about minor typos and excessive functions.

 Squash the second commit! :)

File: packages/react/src/components/button/Button.stories.tsx [hunk 5]
```diff
@@ -3,13 +3,20 @@ import { action } from '@storybook/addon-actions';
 import { ArgsTable, Stories, Title } from '@storybook/addon-docs/blocks';
 
 import { IconShare, IconAngleRight, IconFaceSmile, IconTrash } from '../../icons';
-import { Button } from './Button';
+import { Button, ButtonTheme, ButtonSize, ButtonPresetTheme, ButtonVariant } from './Button';
 
 const onClick = action('button-click');
 
+const getLabel = (label = 'Button'): string => label;
+
```
[the remark this answers is not in the document]
Comment (mrTuomoK) [reply packages/react/src/components/button/Button.stories.tsx:11]: Yeah the function was there already, so not mine  😆

File: CHANGELOG.md (hunk 3 above)
Comment (mrTuomoK) [reply CHANGELOG.md:64]: Good point, missed it!

File: packages/react/src/components/button/Button.tsx (hunk 4 above)
Comment (mrTuomoK) [reply packages/react/src/components/button/Button.tsx:177]: This I do not know? Let's discuss in daily perhaps

**Comment by SillanmakiA** [comment]
> Padding seems weird in these react storybook button examples:
> 
> * with icons
> * Loading
> * Small
> 
> Also buttons in datepicker and accordion have odd paddings on buttons when focused.
> 
> In Core storybook buttons look normal.

Can you elaborate a bit on what you found weird? The buttons seem to be the same between react and core and as designed.

**Comment by Riippi** [comment]
> > Padding seems weird in these react storybook button examples:
> > 
> > * with icons
> > * Loading
> > * Small
> > 
> > Also buttons in datepicker and accordion have odd paddings on buttons when focused.
> > In Core storybook buttons look normal.
> 
> Hmm. Everything is according to designs, could you @SillanmakiA check them (check also Accordion and DatePicker)? @Riippi the Button paddings and designs have changed a bit, but we'll check everything again. Oh the loki-test images take some stupid empty space in themselves from some container around the buttons at least (so those are not any paddings or other in the components themselves)

Maybe there is some issue on my machine if things look ok for others?

Screenshot 2024-04-18 at 14 37 20
Screenshot 2024-04-18 at 14 38 13
Screenshot 2024-04-18 at 14 37 48

**Comment by Riippi** [comment]
> > Padding seems weird in these react storybook button examples:
> > 
> > * with icons
> > * Loading
> > * Small
> > 
> > Also buttons in datepicker and accordion have odd paddings on buttons when focused.
> > In Core storybook buttons look normal.
> 
> Can you elaborate a bit on what you found weird? The buttons seem to be the same between react and core and as designed.

Replied on Tuomo's message about this.

**Review by Riippi — APPROVED** [review]
👍

File: packages/react/src/components/button/Button.stories.tsx [hunk 6]
```diff
@@ -53,29 +92,89 @@ FullWidth.storyName = 'Full width';
...
+      Custom
+    </Button>
+  );
+};
+
 export const Loading = () => (
-  <Button isLoading loadingText="Saving your changes">
-    Button
+  <Button iconStart={<LoadingSpinner small />} variant={ButtonVariant.Loading}>
+    Saving your changes
```
Comment (NikoHelle) [inline packages/react/src/components/button/Button.stories.tsx:177]: All good! But... I'd consider naming the "Loading"-variant to something else like "Clear", "Plain", "Text" etc.

Comment (mrTuomoK) [reply packages/react/src/components/button/Button.stories.tsx:177]: Hmm... btw we could even get rid of setting the iconStart by using the variant? But then it should be loading-something? @NikoHelle ?

**Review by NikoHelle — APPROVED** [review]
I'll leave padding etc verifications to designers. Otherwise looks very good.

**Pull request merged**